File: code/sdk/android/cloudsheet-Android-sdk/src/main/kotlin/com/opencloudsheet/storage/workbook/delete/OneDrivePersonalWorkbookDeleter.kt
package com.opencloudsheet.storage.workbook.delete

import android.util.Log
import com.opencloudsheet.auth.IAuthenticator
import com.opencloudsheet.constants.OneDriveClient
import com.opencloudsheet.constants.OneDriveConstants
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Request

/**
 * Deletes workbooks from the current user's personal OneDrive.
 *
 * Uses the Microsoft Graph API to delete workbooks from the authenticated user's drive.
 * This differs from other implementations that may operate on different users' drives.
 *
 * @see <a href="https://learn.microsoft.com/en-us/graph/api/driveitem-delete?view=graph-rest-1.0&tabs=http">Microsoft Graph: Delete Drive Item</a>
 */
class OneDrivePersonalWorkbookDeleter(
    private val authenticator: IAuthenticator,
    private val oneDriveClient: OneDriveClient
) : IDeleteWorkBook {

    companion object {
        private const val TAG = "OneDrivePersonalWorkbookDeleter"
    }

    override suspend fun deleteWorkBook(ownerId: String, fileId: String) {
        val authToken = authenticator.getAuthToken()
            ?: throw IllegalStateException("No authentication token available")

        val deleteUrl = "${OneDriveConstants.BASE_MS_GRAPH_URL}/users('$ownerId')/drive/items('$fileId')"

        Log.d(TAG, "Deleting workbook file: $fileId")

        val request = Request.Builder()
            .url(deleteUrl)
            .addHeader("Authorization", "Bearer $authToken")
            .delete()
            .build()

        withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    val errorBody = response.body?.string() ?: ""
                    Log.e(TAG, "Failed to delete workbook: ${response.code} - ${response.message} - $errorBody")
                    throw Exception("Failed to delete workbook file: HTTP ${response.code}")
                }
                Log.d(TAG, "Workbook file deleted successfully: $fileId")
            }
        }
    }
}

File: code/sdk/ios/OpenCloudSheet/Sources/OpenCloudSheet/Storage/Workbook/Delete/OneDrivePersonalWorkbookDeleter.swift
//
//  OneDrivePersonalWorkbookDeleter.swift
//  OpenCloudSheet
//
//  Deletes workbooks from the current user's personal OneDrive
//

import Foundation

class OneDrivePersonalWorkbookDeleter: IDeleteWorkBook {
    private let authenticator: IAuthenticator
    private let oneDriveClient: OneDriveClient

    init(authenticator: IAuthenticator, oneDriveClient: OneDriveClient) {
        self.authenticator = authenticator
        self.oneDriveClient = oneDriveClient
    }

    func deleteWorkBook(ownerId: String, fileId: String) async throws {
        guard let authToken = try await authenticator.getAuthToken() else {
            throw NSError(domain: "OneDrivePersonalWorkbookDeleter", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let deleteUrl = "\(OneDriveConstants.baseMSGraphURL)/users('\(ownerId)')/drive/items('\(fileId)')"

        print("Deleting workbook file: \(fileId)")

        let url = URL(string: deleteUrl)!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.addValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await oneDriveClient.session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 204 else {
            throw NSError(domain: "OneDrivePersonalWorkbookDeleter", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to delete workbook file"])
        }

        print("Workbook file deleted successfully: \(fileId)")
    }
}
